chrome/content/zotero/xpcom/integration.js

```diff
@@ -869,11 +886,37 @@ Zotero.Integration.Fields.prototype._processFields = Zotero.Promise.coroutine(fu
  */
 Zotero.Integration.Fields.prototype.updateDocument = Zotero.Promise.coroutine(function* (forceCitations, forceBibliography,
 		ignoreCitationChanges) {
-	// Iterate through citations, yielding for UI updates
-	yield Zotero.Promise.each(this._session._updateCitations(), () => {});
+	var startTime = (new Date()).getTime();
 	
-	yield Zotero.Promise.each(
-		this._updateDocument(forceCitations, forceBibliography, ignoreCitationChanges), () => {});
+	// Iterate through citations, yielding for UI updates
+	for (let t of this._session._updateCitations()) {
+		// Promise.each doesn't yield for UI updates here for some reason, so we yield into the next event loop
+		var deferred = Zotero.Promise.defer();
+		setTimeout(deferred.resolve.bind(deferred));
+		yield deferred.promise;
+	}
+	for (let t of this._updateDocument(forceCitations, forceBibliography, ignoreCitationChanges)) {
+		// Promise.each doesn't yield for UI updates here for some reason, so we yield into the next event loop
+		var deferred = Zotero.Promise.defer();
+		setTimeout(deferred.resolve.bind(deferred));
+		yield deferred.promise;
+	}
```

Comment:
Zotero.Promise.delay(1) definitely yields to the next event loop, because it is a delay… It's literally a setTimeout() call. There's no reason to use generators here.